crates/typed-store/src/metrics.rs

```diff
@@ -1,12 +1,103 @@
+    pub num_ops: Arc<AtomicU64>,
+    // Timestamp of the last sample
+    pub latest_sample_ts: Arc<AtomicU64>,
+}
+
+#[derive(Debug, Clone)]
+// A struct for sampling based on number of operations or duration.
+// Sampling happens if the duration expires or number of operations
+// is more than the threshold
+pub struct SamplingInterval {
+    // Sample once every time duration
+    pub once_every_duration: Duration,
+    // Sample once every number of operations
+    pub after_num_ops: u64,
+    // Perf sampler for keeping track of previous sample
+    pub perf_sampler: PerfSampler,
+}
+
+impl Default for SamplingInterval {
+    fn default() -> Self {
+        SamplingInterval {
+            once_every_duration: Duration::from_secs(1),
+            after_num_ops: u64::MAX,
+            perf_sampler: PerfSampler {
+                num_ops: Arc::new(AtomicU64::new(0)),
+                latest_sample_ts: Arc::new(AtomicU64::new(0)),
+            },
+        }
+    }
+}
+
+impl SamplingInterval {
+    pub fn new(once_every_duration: Duration, after_num_ops: u64) -> Self {
+        SamplingInterval {
+            once_every_duration,
+            after_num_ops,
+            perf_sampler: PerfSampler {
+                num_ops: Arc::new(AtomicU64::new(0)),
+                latest_sample_ts: Arc::new(AtomicU64::new(0)),
+            },
+        }
+    }
+    pub fn sample(&self) -> bool {
+        if self.after_num_ops == u64::MAX && self.once_every_duration == Duration::MAX {
+            return false;
+        }
+        if self.after_num_ops != u64::MAX {
+            let num_ops = self.perf_sampler.num_ops.fetch_add(1, Ordering::Relaxed);
+            if num_ops >= self.after_num_ops {
+                if num_ops > self.after_num_ops {
+                    // Multiple threads got here at the same time, only one
+                    // should be able to get the sample
+                    return false;
+                }
+                let now = SystemTime::now()
+                    .duration_since(UNIX_EPOCH)
+                    .expect("create timestamp in timing");
+                self.perf_sampler.num_ops.store(0, Ordering::SeqCst);
+                self.perf_sampler
+                    .latest_sample_ts
+                    .store(now.as_secs(), Ordering::SeqCst);
+                return true;
+            }
+        }
+        if self.once_every_duration == Duration::MAX {
+            return false;
+        }
+        let now = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .expect("create timestamp in timing");
+        let prev_ts = self.perf_sampler.latest_sample_ts.load(Ordering::Relaxed);
+        let ts_since_last_op = now.as_secs() - prev_ts;
+        if ts_since_last_op >= self.once_every_duration.as_secs()
+            && self
+                .perf_sampler
+                .latest_sample_ts
+                .compare_exchange_weak(prev_ts, now.as_secs(), Ordering::SeqCst, Ordering::Relaxed)
```

Comment:
I think it's possible to combine the two branches you have here.   I mean, the check on number of ops elapsed and time leased, you could check both, do the compare and swap, and reset the things all at once in the same branch.  You could use the compare and swap to determine who wins.

Reply:
The reason for doing one at a time is to optimize away the checking of the other one